### ISSUE TITLE
ConcurrentDictionary: Count property is too expensive

### ISSUE THREAD

**Issue by ikopylov** [dotnet/corefx#3357]
In the current implementation of ```Count``` property from ```ConcurrentDictionary<TKey, TValue>``` the full lock is acquired, which is quite expensive. It is much better to add a private field to track element count in the whole dictionary and update it by ```Interlocked``` operations.

It would be great if someone will implement this performance tweak.

**Comment by stephentoub** [comment]
>  It is much better to add a private field to track element count in the whole dictionary and update it by Interlocked operations

It's not, though.  When many threads are mutating the dictionary, those interlocked operations on a single shared field are going to cause contention across all of those threads, slowing down adds/removes/updates/etc; the design of ConcurrentDictionary is such that different threads typically end up taking different locks, such that this contention is eliminated, and we chose to prioritize such operations rather than the typically comparatively rare act of accessing the Count.

**Comment by VSadov** [comment]
One extra Interlocked operation on add would be noticeable on the performance of Concurrent dictionary. In particular if many threads are adding since they would share the counter.

It is possible to mitigate the sharing by using a striped counter, but is this an important enough scenario to bother?

Count of a concurrent table would be useful only if caller can be sure the table is in quiescent state and the lock should be fairly cheap in such scenario.

**Comment by JonHanna** [comment]
Interlocked on a single field from multiple cores would make for a bottleneck as every thread competes to update it. If you are going to maintain a tally, you would really want to do it with an array of ints, and add them when the count is requested. They'd also need to be spaced from each other in memory to avoid false sharing, so it would need to cover several cache lines (say maybe covering 1KiB to spread it out across 16 cache lines on chips with 64-byte cache lines).

**Comment by JonHanna** [comment]
I've implemented a concurrent dictionary which takes a completely different approach to that in the framework. It needed to maintain a count as part of its implementation, and if I replace its striped counter with a single integer it becomes massively slower, even if I don't use interlocked (since it only really needs an estimate and could tolerate the occasional mis-write). It really does make a very large difference in practice.

**Comment by sharwell** [comment]
Right now I'm mostly interested in the reason why `Count` can't be implemented as follows:

```csharp
public int Count
{
    get
    {
        int count = 0;

        // Compute the count, we allow overflow
        int countPerLock = _tables._countPerLock;
        for (int i = 0; i < countPerLock.Length; i++)
        {
            count += countPerLock[i];
        }

        return count;
    }
}
```

**Comment by jasonwilliams200OK** [comment]
@CodesInChaos shared some useful thoughts on a related topic here:
http://stackoverflow.com/a/5222021.

**Comment by stephentoub** [comment]
> Right now I'm mostly interested in the reason why  Count  can't be implemented as follows

Because it was designed to have snapshot semantics.

**Comment by sharwell** [comment]
> Implemented as such, you could even end up with a higher count than was ever in the collection.

Thanks, this was enough to figure out an example ordering:

Thread 1:

* Read `Count`

Thread 2:

1. Add *X*
1. Remove *X*
1. Add *Y*

The count operation could read the `countPerLock` bucket holding *X* between steps 1 and 2, followed by reading the `countPerLock` bucket holding *Y* after step 3. The count would be reported as though both *X* and *Y* were in the collection, even though there is no point in time where this could be true.

**Comment by ikopylov** [comment]
My scenario was to collect statistics on the number of elements in the system (i.e. in the ```ConcurrentDictionary``` containing all elements). That statistic are updated after every change according to the ```Count``` property. After performance profiling I was very surprised that ```Count``` property took 20% of working time. So I've ended up with special field which updated by ```Interlocked``` operations.

> we chose to prioritize such operations rather than the typically comparatively rare act of accessing the Count.

It is very unexpected that accessing to ```Count``` property is so slow. Probably, this should be stated in XML documentation to that property.

> if I replace its striped counter with a single integer it becomes massively slower

And what was the hardware configuration? Was it a single multi-core CPU or multi-CPU system?
In my experience, I've never encountered critical performance problem because of the contention on a single multi-core CPU.

> Because it was designed to have snapshot semantics. 

Is it really important to use snapshot semantic for ```Count``` property of concurrent collection? In any case, one should not rely on the value of this property, because it can be changed at any moment of time.

I really like the solution without any lock, proposed by @sharwell.

> Implemented as such, you could even end up with a higher count than was ever in the collection.

Again, Is it really a probelm for concurrent collection?

**Comment by JonHanna** [comment]
> And what was the hardware configuration? Was it a single multi-core CPU or multi-CPU system?

Single Core i7.

> In my experience, I've never encountered critical performance problem because of the contention on a single multi-core CPU.

The different cores can certainly have false sharing between them. Even if they didn't, what about the multi-CPU setups? Adding more cores should not make things get seriously slower.

**Comment by stephentoub** [comment]
> Is it really a probelm for concurrent collection?

That it could return an answer that was never true?  We thought so, yes.

FWIW, at this point the property is also documented to behave this way: "This property has snapshot semantics and represents the number of items in the ```System.Collections.Concurrent.ConcurrentDictionary<TKey, TValue>``` at the moment when the property was accessed."

> I really like the solution without any lock, proposed by @sharwell.

The summing of the counts per bucket is exactly what the current implementation does; the difference is that it takes the locks before and releases them after in order to get an accurate count.

If you're asking for ConcurrentDictionary to expose an ApproximateCount property or something like that, it's something that could be considered, but I don't think the Count property should be changed, unless there's a way to maintain it returning an accurate count without impacting the performance of other operations on the collection.

**Comment by JonHanna** [comment]
I certainly agree that it can be useful to have an estimate of a count without a snapshot, in high contention the current count is out of date the moment it returns and releases its locks anyway, and for some uses an estimate that didn't hit the locks would be just as useful as a precise count of how things were at some recent point in the past.

**Comment by omariom** [comment]
It would be better if it had IsEmpty and DebugCount properties and ICollection.Count implemented explicitly.

**Comment by sharwell** [comment]
I'm actually in favor of the original proposal (using Interlocked to maintain the count). I believe that on the majority of client PCs which will be using this type, and in the majority of applications which are using this type, the performance overhead of increased contention over this field would not outweigh the gains made by improving the performance of `Count`.

**Comment by stephentoub** [comment]
> I believe that on the majority of client PCs which will be using this type, and in the majority of applications which are using this type, the performance overhead of increased contention over this field would not outweigh the gains made by improving the performance of Count.

This assumes a particular usage pattern of Count.  In the majority of uses I've seen of the type, especially server-side, Count is rarely if ever used.

> I would love to run performance evaluations to see exactly how much overhead this would have in simulated scenarios.

On my quad-core laptop, with a simple test that just repeatedly adds/removes from the collection, a single-threaded test that incurs an extra interlocked per add or remove decreases throughput by ~5-7%.  When I change it to bang on the collection in parallel with four threads, that increases to an ~15-20% reduction.  That's just a quick test on my machine, though; I'd expect the impact to be more sever with more cores, with multiple sockets, etc.

**Comment by ikopylov** [comment]
> Even if they didn't, what about the multi-CPU setups? Adding more cores should not make things get seriously slower.

Maintaining cache coherence across CPU is slower than across cores inside a single CPU. So contention and false sharing effects should give a higher performance hit in multi-socket setups.

> If you're asking for ConcurrentDictionary to expose an ApproximateCount property or something like that, it's something that could be considered

If it is required to preserve snapshot semantics then this will be the best solution. In my opinion, the fast approximate Count is more useful in practice than precise Count.

> FWIW, at this point the property is also documented to behave this way: "This property has snapshot semantics and represents the number of items in the ```System.Collections.Concurrent.ConcurrentDictionary<TKey, TValue>``` at the moment when the property was accessed."

But it says nothing about the potential performance degradation in the high concurrency scenarios.

> When I change it to bang on the collection in parallel with four threads, that increases to an ~15-20% reduction.

Ok, it is not a huge, but still noticeable reduction.


I think that additional ```ApproximateCount``` property would be fine.

**Comment by stephentoub** [comment]
Given the discussion, I'm going to close this issue.  If you need an additional member added, please open a separate issue to propose it.  Thanks.
